### PR TITLE
[Xamarin.Android.Build.Tasks] First Pass at adding Unit tests for the MSBuild Tasks

### PR DESCRIPTION
**Pull request by dellis1972**
This commit adds basic support for the MSBuild Unit Tests. The goal
of these tests is to ensure that the build of an android applicaition
works consistently on all the supported platforms.

The basic layout of a Unit test is as follows

	[Test]
	public void BuildReleaseApplication ()
	{
		var proj = new XamarinAndroidApplicationProject () {
			IsRelease = true,
		};
		using (var b = CreateApkBuilder (Path.Combine ("temp", TestContext.CurrentContext.Test.Name))) {
			Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
		}
	}

It is a standard Unit test. First we create a XamarinAndroidApplicatonProject
and set its properties. You can use the proj to add new source files,
references, assets, resources and nuget packages. By default you will get
a standard "HelloWorld" android application.
Once you have the project you will need to create either a Apk or Dll builder
via the helper methods

	CreateApkBuilder
	CreateDllBuilder

CreateApkBuilder will create an apk and by default will execute the
`SignAndroidPackage` build target. the CreateDllBuilder will produce
a dll. The source files are created in a temp directory relative to
the build output of the unit tests. By default this is

	src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.Android.Build.Tests/bin/$(Configuraiton)/temp

Once you have a builder you can then call Build passing in the project. There are
also methods available for Clean and Save. Running any of these will cause
the Unit test to shell out to xbuild/msbuild and attempt to build the project.
Test results are written to a build.log file.
If a unit test passes then the test directory is removed. If a test fails
the directory and the build.log will be left in place for investigation.

### CODE REVIEW

**Comment by dellis1972** [comment]
@jonpryor don't merge this yet (if you are tempted) I still need to update it to use the xabuild script rather than system xbuild. Also it needs updating to look for the built framework.

**Comment by jonpryor** [comment]
+++ src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.Android.Build.Tests/BuildTest.cs

Convention is that unit tests go into a nested [`Test` directory](https://github.com/mono/mono/tree/master/mcs/class/corlib/Test), or [`Tests`](https://github.com/xamarin/java.interop/tree/master/src/Java.Interop/Tests) (oops), not `UnitTests`, i.e. this should be `src/Xamarin.Android.Build.Tasks/Test/Xamarin.Android.Build.Tests/BuildTest.cs`.

File: src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.ProjectTools/Xamarin.ProjectTools.csproj [hunk 1]
```diff
@@ -0,0 +1,115 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+	<Import Project="..\..\..\..\Configuration.props" />
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.ProjectTools/Xamarin.ProjectTools.csproj]: I wonder why this file is indented with tabs while the other `.csproj` files are indented with spaces, two spaces per indent, e.g. [Mono.Android.csproj](https://github.com/xamarin/xamarin-android/blob/c7a41e5/src/Mono.Android/Mono.Android.csproj).

Should it be fixed for consistency?

File: src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.ProjectTools/Xamarin.ProjectTools.csproj [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		<OutputType>Library</OutputType>
+		<RootNamespace>Xamarin.ProjectTools</RootNamespace>
+		<AssemblyName>Xamarin.ProjectTools</AssemblyName>
+		<TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+	</PropertyGroup>
+	<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+		<DebugSymbols>true</DebugSymbols>
+		<DebugType>full</DebugType>
+		<Optimize>false</Optimize>
+		<OutputPath>bin\Debug</OutputPath>
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/UnitTests/Xamarin.ProjectTools/Xamarin.ProjectTools.csproj]: We should consider putting this into `$(topdir)/bin/Test$(Configuration)`, i.e. `..\..\..\bin\TestDebug`. This is what the Java.Interop repo does.

**Comment by jonpryor** [comment]
This also needs a way to *run* the unit tests, e.g. a `make run-all-tests` target, a'la [Java.Interop](https://github.com/xamarin/java.interop/blob/8b647c2/Makefile#L120-L131).

**Comment by dellis1972** [comment]
@jonpryor I've updated this to move the UnitTests to Tests and added a runner.

File: Makefile [hunk 3]
```diff
@@ -19,5 +25,21 @@ prepare:
...
+# $(call RUN_TEST,filename,log-lref?)
+define RUN_TEST
+	MONO_TRACE_LISTENER=Console.Out \
+	$(RUNTIME) --runtime=v4.0.0 \
+		$(NUNIT_CONSOLE) $(NUNIT_EXTRA) $(1) \
+		$(if $(RUN),-run:$(RUN)) \
+		-output=bin/Test$(CONFIGURATION)/TestOutput-$(basename $(notdir $(1))).txt ;
+endef
+
+run-tests: $(TESTS)
```
Comment (jonpryor) [inline Makefile]: This should be `run-nunit-tests`, to differentiate from the `run-apk-tests` in PR #32.

File: Makefile [hunk 4]
```diff
@@ -19,5 +25,21 @@ prepare:
...
 
+
+run-all-tests: run-tests 
+
 clean:
 	$(MSBUILD) /t:Clean
+
+
+# $(call RUN_TEST,filename,log-lref?)
+define RUN_TEST
```
Comment (jonpryor) [inline Makefile]: Likewise, this should be `RUN_NUNIT_TEST`.

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/BaseTest.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+				HasDevices = output.Split ('\n').Where (s => s.Contains ("emulator-")).Count () > 0;
+			}
+			catch (Exception ex) {
+				Console.Error.WriteLine ("Failed to determine whether there is Android target emulator or not" + ex);
+			}
+		}
+
+		public static readonly bool HasDevices;
+
+		public enum LicenceType
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/BaseTest.cs]: Do we need this enum?

File: Makefile [hunk 6]
```diff
@@ -1,6 +1,12 @@
 V             ?= 0
 CONFIGURATION = Debug
 MSBUILD       = xbuild /p:Configuration=$(CONFIGURATION) $(MSBUILD_ARGS)
+RUNTIME       := $(shell if [ -f `which mono64` ] ; then echo mono64 ; else echo mono; fi) --debug=casts
+
+TESTS = \
```
Comment (jonpryor) [inline Makefile]: Sorry; this should likewise be `NUNIT_TESTS`.

**Pull request merged**